google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/stub/metrics/BuiltinMetricsTracerTest.java

```diff
@@ -0,0 +1,341 @@
+    assertThat(longValue.getValue()).isEqualTo(0);
+  }
+
+  @Test
+  public void testReadRowsApplicationLatency() throws Exception {
+    final long applicationLatency = 1000;
+    final SettableApiFuture future = SettableApiFuture.create();
+    final AtomicInteger counter = new AtomicInteger(0);
+    // We want to measure how long application waited before requesting another message after
+    // the previous message is returned from the server. Using ResponseObserver here so that the
+    // flow will be
+    // onResponse() -> sleep -> onRequest() (for the next message) which is exactly what we want to
+    // measure for
+    // application latency.
+    // If we do readRowsCallable().call(Query.create(TABLE_ID)).iterator() and iterate through the
+    // iterator and sleep in
+    // between responses, when the call started, the client will pre-fetch the first response, which
+    // won't be counted
+    // in application latency. So the test will be flaky and hard to debug.
+    stub.readRowsCallable()
+        .call(
+            Query.create(TABLE_ID),
+            new ResponseObserver<Row>() {
+              @Override
+              public void onStart(StreamController streamController) {}
+
+              @Override
+              public void onResponse(Row row) {
+                try {
+                  counter.incrementAndGet();
+                  Thread.sleep(applicationLatency);
+                } catch (InterruptedException e) {
+                }
+              }
+
+              @Override
+              public void onError(Throwable throwable) {
+                future.setException(throwable);
+              }
+
+              @Override
+              public void onComplete() {
+                future.set(null);
+              }
+            });
+    future.get();
+
+    verify(statsRecorderWrapper).putApplicationLatencies(longValue.capture());
+
+    assertThat(counter.get()).isGreaterThan(0);
+    assertThat(longValue.getValue()).isAtLeast(applicationLatency * counter.get());
+    assertThat(longValue.getValue()).isLessThan(applicationLatency * (counter.get() + 1));
+  }
+
+  @Test
+  public void testRetryCount() {
+    stub.mutateRowCallable()
+        .call(RowMutation.create(TABLE_ID, "random-row").setCell("cf", "q", "value"));
+
+    verify(statsRecorderWrapper).putRetryCount(intValue.capture());
+
+    assertThat(intValue.getValue()).isEqualTo(3);
+  }
+
+  @Test
+  public void testMutateRowAttempts() {
+    stub.mutateRowCallable()
+        .call(RowMutation.create(TABLE_ID, "random-row").setCell("cf", "q", "value"));
+
+    verify(statsRecorderWrapper, times(4))
+        .record(status.capture(), tableId.capture(), zone.capture(), cluster.capture());
+    assertThat(zone.getAllValues().get(0)).isEqualTo(UNDEFINED);
+    assertThat(zone.getAllValues().get(1)).isEqualTo(UNDEFINED);
+    assertThat(zone.getAllValues().get(2)).isEqualTo(UNDEFINED);
```

Comment:
here and everywhere else:
`assertThat(zone.getAllValues().containsExactly(UNDEFINED, UNDEFINED, UNDEFINED).inOrder()`